### PR TITLE
Add example script to produce true and reco bg cube model

### PR DESCRIPTION
**Pull request by mapazarr**
The script uses the same model (except for absolute normalization) to produce a true cube bg model and a reco cube bg model. The models can be used to test the cube bg model production and can be compared to each other using the plot_bg_cube_model_comparison.py example script.

### CODE REVIEW

**Comment by mapazarr** [comment]
Und noch einen Kaffee! :-)
Mergen?

**Comment by cdeil** [comment]
@mapazarr – Thanks for adding this!

As with your other recent PRs I'm OK with merging this without much review, but still I'd like to do some review to avoid accumulating more and more [technical debt](https://en.wikipedia.org/wiki/Technical_debt) in Gammapy, i.e. things that need to be reviewed / tested / cleaned up later.

**Comment by cdeil** [comment]
Please don't use minus `-` in Python file names, use underscore `_` instead.

File: examples/make_bg_cube_models_true-reco_test.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    bg_cube_model.write('{}_image.fits.gz'.format(outfile),
+                        format='image', clobber=overwrite)
+
+
+if __name__ == '__main__':
+
+    outdir = OUTDIR
+    overwrite = OVERWRITE
+
+    # create output folder
```
Comment (cdeil) [inline examples/make_bg_cube_models_true-reco_test.py]: These ~10 lines of code have appeared now a few times in your code I think.
Please make it a utility function.

File: gammapy/background/cube.py [hunk 2]
```diff
@@ -809,6 +809,38 @@ def plot_spectrum(self, coord, ax=None, style_kwargs=None):
             plt.close(fig)
         return ax
 
+    @property
+    def integral(self):
+        """Integral of the cube (`~astropy.units.Quantity`)"""
```
Comment (cdeil) [inline gammapy/background/cube.py]: What is the "Integral of the cube"?

I think it's predicted counts?

If so you could call it `npred` or `npred_total` (see 
https://gammapy.readthedocs.org/en/latest/api/gammapy.data.SpectralCube.html#gammapy.data.SpectralCube.integral_flux_image and
https://gammapy.readthedocs.org/en/latest/api/gammapy.data.compute_npred_cube.html),
or at least mention this in the docstring.

Also maybe add "scalar" to make it clearer?
```
(scalar `~astropy.units.Quantity`)
```

File: gammapy/datasets/make.py [hunk 3]
```diff
@@ -83,32 +83,55 @@ def sigma_energy_theta(energy, theta, sigma):
 
 
 def make_test_observation_table(observatory_name='HESS', n_obs=10,
-                                datestart=None, dateend=None,
+                                az_range=Angle([0, 360], 'degree'),
+                                alt_range=Angle([45, 90], 'degree'),
+                                date_range=(Time('2010-01-01T00:00:00',
+                                                 format='isot', scale='utc'),
```
Comment (cdeil) [inline gammapy/datasets/make.py]: Remove `format='isot', scale='utc'` ... UTC is the default and the format it's important.

File: gammapy/background/cube.py [hunk 4]
```diff
@@ -809,6 +809,38 @@ def plot_spectrum(self, coord, ax=None, style_kwargs=None):
...
+        integral = self.data*bin_volume
+
+        return integral.sum()
+
+    @property
+    def integral_images(self):
+        """Integral of the cube images (`~astropy.units.Quantity`)
+
+        Calculate the integral of each energy bin (slice) in the
+        cube. Returns an array of integrals.
```
Comment (cdeil) [inline gammapy/background/cube.py]: Again, from the method name and docstring it's not obvious to me what this does.

It looks to me like this is computing a 1-dim `npred` array for each energy bin?

File: gammapy/datasets/make.py [hunk 5]
```diff
@@ -83,32 +83,55 @@ def sigma_energy_theta(energy, theta, sigma):
...
         Name of the observatory; a list of choices is given in
         `~gammapy.obs.observatory_locations`.
-    n_obs : int
+    n_obs : int, optional
         Number of observations for the obs table.
-    datestart : `~astropy.time.Time`, optional
-        Starting date for random generation of observation start time.
-    dateend : `~astropy.time.Time`, optional
-        Ending date for random generation of observation start time.
+    az_range : `~astropy.coordinatesAngle`, optional
```
Comment (cdeil) [inline gammapy/datasets/make.py]: Missing dot: `~astropy.coordinatesAngle` -> `~astropy.coordinates.Angle`

File: gammapy/datasets/make.py [hunk 6]
```diff
@@ -83,32 +83,55 @@ def sigma_energy_theta(energy, theta, sigma):
...
+        Azimuth angle range (start, end) for random generation of
+        observation pointing positions.
+    alt_range : `~astropy.coordinatesAngle`, optional
```
Comment (cdeil) [inline gammapy/datasets/make.py]: Missing dot again.

File: gammapy/datasets/make.py [hunk 7]
```diff
@@ -392,7 +411,13 @@ def make_test_bg_cube_model(detx_range=Angle([-10., 10.], 'degree'),
 
 def make_test_dataset(fits_path, overwrite=False,
                       observatory_name='HESS', n_obs=10,
-                      datestart=None, dateend=None,
+                      az_range=Angle([0, 360], 'degree'),
+                      alt_range=Angle([45, 90], 'degree'),
+                      date_range=(Time('2010-01-01T00:00:00',
```
Comment (cdeil) [inline gammapy/datasets/make.py]: Simplify:
```python
date_range=Time(['2010-01-01', '2015-01-01'])
```

File: gammapy/datasets/make.py [hunk 8]
```diff
@@ -424,10 +449,18 @@ def make_test_dataset(fits_path, overwrite=False,
         `~gammapy.obs.observatory_locations`.
     n_obs : int
         Number of observations for the obs table.
-    datestart : `~astropy.time.Time`, optional
-        Starting date for random generation of observation start time.
-    dateend : `~astropy.time.Time`, optional
-        Ending date for random generation of observation start time.
+    az_range : `~astropy.coordinatesAngle`, optional
```
Comment (cdeil) [inline gammapy/datasets/make.py]: Again: missing docs (also two lines down)

File: gammapy/scripts/make_bg_cube_models.py [hunk 9]
```diff
@@ -311,8 +311,7 @@ def stack_observations(fits_path, outdir, overwrite, method='default'):
         bg_cube_model = make_bg_cube_model(observation_table, fits_path, method)
 
         # save model to file
-        outfile = outdir +\
-                 '/bg_cube_model_group{}'.format(group)
+        outfile = outdir + '/bg_cube_model_group{}'.format(group)
```
Comment (cdeil) [inline gammapy/scripts/make_bg_cube_models.py]: Use `os.path.join` ... this doesn't work on Windows. :smile:

File: examples/make_bg_cube_models_true-reco_test.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    METHOD = 'default'
+
+    fits_path = 'test_dataset'
+    overwrite = OVERWRITE
+    test = TEST
+    group_id = GROUP_ID
+
+    # create output folder
+    outdir = OUTDIR + '/reco'
+    if not os.path.isdir(outdir):
```
Comment (cdeil) [inline examples/make_bg_cube_models_true-reco_test.py]: Again the same code to make a folder ... use utility function.

File: CHANGES.rst [hunk 11]
```diff
@@ -24,14 +24,16 @@ For plans and progress see https://github.com/gammapy/gammapy/milestones/0.4
...
 
 Pull requests
 +++++++++++++
 
 - Make background cube models [#319] (Manuel Paz Arribas)
-- Production of true/reco bg cube models should use the same model. [#335] (Manuel Paz Arribas)
-- Fix TS map boundary handling [#332] (Axel Donath) 
+-  Production of true/reco bg cube models should use the same model. [#335] (Manuel Paz Arribas)
+- Fix TS map boundary handling [#332] (Axel Donath)
+- Added example script to produce true and reco bg cube model. [#338] (Manuel Paz Arribas)
```
Comment (cdeil) [inline CHANGES.rst]: Let's use a uniform format for GH issue titles and changelog. Active voice ("Add" instead of "Added") and no dot at the end of the line.

File: examples/make_bg_cube_models_true-reco_test.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+    return obs_groups
+
+
+def make_true_model():
+    """Make a true bg cube model."""
+
+    overwrite = OVERWRITE
+
+    # create output folder
```
Comment (cdeil) [inline examples/make_bg_cube_models_true-reco_test.py]: Use utility function.

File: examples/make_bg_cube_models_true-reco_test.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+ALT_RANGE = Angle([72, 90], 'degree')
+GROUP_ID = 27
+OUTDIR = 'bg_cube_models'
+OVERWRITE = False
+
+
+def create_dummy_observation_grouping():
+    """Define dummy observation grouping."""
+
+    altitude_edges = ALT_RANGE
```
Comment (cdeil) [inline examples/make_bg_cube_models_true-reco_test.py]: I think this function can be shortened by half:
```python
def create_dummy_observation_grouping():
    alt_ax = ObservationGroupAxis('ALT', ALT_RANGE, 'bin_edges')
    az_ax = ObservationGroupAxis('AZ', AZ_RANGE, 'bin_edges')
    obs_groups = ObservationGroups([alt_ax, az_ax])
    obs_groups.obs_groups_table['GROUP_ID'][0] = GROUP_ID

    return obs_groups
```

The following functions can also be simplified / improved.

Just in case you say "this is just an example and not worth the effort of code review / polishing", I'd say "especially for examples it's important to make them nice and clean" ... it's how users / co-developers learn and what determines whether they like Gammapy or now.

File: examples/make_bg_cube_models_true-reco_test.py [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+
+    # create dummy observation grouping
+    obs_groups = create_dummy_observation_grouping()
+    # save
+    outfile = outdir + '/bg_observation_groups.ecsv'
+    print('Writing {}'.format(outfile, overwrite=overwrite))
+    obs_groups.write(outfile)
+
+    # use binning from CubeBackgroundModel.define_cube_binning
+    detx_range = (Angle(-0.07, 'radian').to('degree'),
```
Comment (cdeil) [inline examples/make_bg_cube_models_true-reco_test.py]: If you're using the same binning for x and y, just have one `det_range` variable and use it twice in the call below.

File: examples/plot_bg_cube_model_comparison.py [hunk 15]
```diff
@@ -17,11 +17,18 @@
 GRAPH_DEBUG = 0
 SAVE = 0
 
+NORMALIZE = 0 # normalize 1 w.r.t. 2 (i.e. true w.r.t. reco)
+              # 0: do not normalize
+              # 1: normalize w.r.t. cube integral
+              # 2: normalize w.r.t images integral (normalize each image on its own)
+
 input_dir1 = '/home/mapaz/astropy/working_dir/gammapy_scripts/20150819_ready_to_merge_PR319/bg_cube_models_gammapy_a_la_michi'
```
Comment (cdeil) [inline examples/plot_bg_cube_model_comparison.py]: Hard-coded PATHs make it hard and unlikely that others will run the example.
So it would be much better if you make this a command line option.
For quick scripts it's OK to just use `input_dir = sys.argv[1]` ... no need to use an `ArgumentParser`, so it's just one or two extra lines to change this.

File: gammapy/background/cube.py (hunk 2 above)
Comment (mapazarr) [inline gammapy/background/cube.py]: It depends on the contents of the cube. If it is counts, it is a scalar, and it gives the total number of counts, but if it is the bg rate it will return the number of counts corrected by the livetime (units of 1/s).

File: gammapy/datasets/make.py (hunk 3 above)
Comment (mapazarr) [inline gammapy/datasets/make.py]: `format='isot', scale='utc'` is the standard we defined in file:///home/mapaz/astropy/development_code/gammapy/docs/_build/html/time/index.html#time-handling-in-gammapy
It might be the default in Astropy right now (I'm not sure). But if it changes at some point for some reason, we would have to change all entries throughout Gammapy, to be consistent with our standard. Isn't it better to just explicitly give the format and scale?

BTW, what do you mean by `it's important'?

File: gammapy/background/cube.py (hunk 2 above)
Comment (cdeil) [inline gammapy/background/cube.py]: OK, then just say that in the docstring, e.g. "the integral returned has the unit of the values in the cube times solid angle times energy". You could even mention the one or two common examples that we have.

File: gammapy/datasets/make.py (hunk 3 above)
Comment (cdeil) [inline gammapy/datasets/make.py]: Let's adopt the Astropy default for Gammapy and trust that they won't change it.
(It's highly unlikely ... it would break a lot of user codes if they did ... satellites falling out of the sky ... doomsday.)

I wanted to write "the format is not important", because it's not used for computations.

**Comment by cdeil** [comment]
@mapazarr – There's a merge conflict (probably in the changelog) now that #337 is merged. Please rebase against master.

**Comment by mapazarr** [comment]
I rebased, but travis-ci didn't started: strange...

File: gammapy/datasets/make.py [hunk 16]
```diff
@@ -566,11 +611,16 @@ def make_test_eventlist(observation_table,
...
-    energy = Quantity((e_min.value - e_max.value) *
-                      random_state.power(a=index, size=n_events) + e_max.value, 'TeV')
+    energy = Quantity(sample_powerlaw(e_min.value, e_max.value,
+                                      spectral_index + 1, # TODO: need to add 1
+                                                          # to the index because
+                                                          # sample_powerlaw
+                                                          # subtracts it, why??!!!
+                                      size=n_events,
+                                      random_state=random_state),
+                      'TeV')
```
Comment (mapazarr) [inline gammapy/datasets/make.py]: @cdeil 
I found a bug: when sampling the energies for the simulated dataset, the spectrum was not a power-law (maybe `numpy.random.RandomState.power` is buggy, or it doesn't work as I thought).

I'm now using `sample_powerlaw`, which seems to work fine, BUT, I have to pass the index as `spectral_index + 1` because the function internally does `exp = 1. - gamma`:
https://github.com/gammapy/gammapy/blob/master/gammapy/utils/random.py#L127
I don't know why does the function need to subtract 1 to the specified index: bug or feature?

Comment (cdeil) [inline gammapy/datasets/make.py]: I'd say bug.

See http://nbviewer.ipython.org/gist/cdeil/3605f1af22d51d0060bb for a check and recommendation what to do (change index by 1 and improve docstring of `sample_powerlaw`).
Can you please do it in this PR or do you want me to do this in a separate PR?

**Comment by cdeil** [comment]
I've seen this before that travis-ci doesn't start.

I think what happens is that Github sends a notification to travis-ci to start a build when you push a change to a branch that's connected to a pull request.

If travis-ci has availability issues at that moment, then this notification is dropped and travis-ci will not poll Github to ask if there have been updates, i.e. travis-ci will never start.

If you push an update to this branch now, and travis-ci is working OK (see https://www.traviscistatus.com/), then travis-ci should start.

**Comment by mapazarr** [comment]
Ok, now true and reco are very similar:
https://hess-confluence.desy.de/confluence/display/HESS/HESS+Open+Source+Tools+-+HOWTO+create+cube+background+models+using+Gammapy#HESSOpenSourceTools-HOWTOcreatecubebackgroundmodelsusingGammapy-Comparisontrue/reco

**Comment by mapazarr** [comment]
@cdeil there seems to be some incompatibility issues between numpy 1.6 or 1.7 and astropy in the travis-ci allowed failure jobs.

**Comment by cdeil** [comment]
I think the best is to finish this PR this week and leave the investigation of the remaining differences true / reco to the future. IMO this is just task number 21 :smile:  in #333 ... I'll try to work on that myself in the coming months or get a student.

**Comment by cdeil** [comment]
> @cdeil there seems to be some incompatibility issues between numpy 1.6 or 1.7 and astropy in the travis-ci allowed failure jobs.

I'm short on time this week, my suggestion would be to just ignore this.
(or file a separate issue if you like.)

File: gammapy/background/cube.py [hunk 17]
```diff
@@ -818,6 +819,47 @@ def divide_bin_volume(self):
...
+        delta_energy = self.energy_edges[1:] - self.energy_edges[:-1]
+        delta_y = self.coordy_edges[1:] - self.coordy_edges[:-1]
+        delta_x = self.coordx_edges[1:] - self.coordx_edges[:-1]
+        # define grid of deltas (i.e. bin widths for each 3D bin)
+        delta_energy, delta_y, delta_x = np.meshgrid(delta_energy, delta_y,
+                                                     delta_x, indexing='ij')
+        if do_not_force_mev_units:
+            bin_volume = delta_energy*(delta_y*delta_x).to('sr')
+        else:
+            bin_volume = delta_energy.to('MeV')*(delta_y*delta_x).to('sr')
```
Comment (mapazarr) [inline gammapy/background/cube.py]: This takes care that MeV units will be used for the bg rates per default. There is though a flag to avoid this if desired (turned off per default).

Comment (mapazarr) [inline gammapy/background/cube.py]: Further discussion on this topic should go here: https://github.com/gammapy/gammapy/commit/e0a3b68f2a2ff762e83660e55a648975e2211fde

File: gammapy/background/make.py [hunk 18]
```diff
@@ -12,7 +12,7 @@
            ]
 
 
-def make_bg_cube_model(observation_table, fits_path, method='default'):
+def make_bg_cube_model(observation_table, fits_path, method='default', do_not_force_mev_units=False):
```
Comment (mapazarr) [inline gammapy/background/make.py]: OK, I moved the enforcing of the units here.

File: gammapy/background/make.py [hunk 19]
```diff
@@ -46,6 +51,11 @@ def make_bg_cube_model(observation_table, fits_path, method='default'):
         Path to the data files.
     method : {'default', 'michi'}, optional
         Bg cube model calculation method to apply.
+    do_not_force_mev_units : bool, optional
+        Set to `True` to use the same energy units as the energy
+        binning for the bg rate. (Only applicable to the *default*
+        method; the *michi* always enforces `MeV` in the bg rate
+        units.)
```
Comment (cdeil) [inline gammapy/background/make.py]: Can you make this an option `unit='1 / (MeV sr s)'` that is applied for any method?
Simpler, more consistent (across methods) and more flexible, no?

File: examples/make_bg_cube_models_true-reco_test.py (hunk 1 above)
Comment (mapazarr) [inline examples/make_bg_cube_models_true-reco_test.py]: Done.

File: gammapy/background/cube.py (hunk 2 above)
Comment (mapazarr) [inline gammapy/background/cube.py]: Done.

File: gammapy/datasets/make.py (hunk 3 above)
Comment (mapazarr) [inline gammapy/datasets/make.py]: Done (everywhere in gammapy).

File: gammapy/background/cube.py (hunk 4 above)
Comment (mapazarr) [inline gammapy/background/cube.py]: Done (similar to the previous one).

File: gammapy/datasets/make.py (hunk 5 above)
Comment (mapazarr) [inline gammapy/datasets/make.py]: Good catch!
Fixed.

File: gammapy/datasets/make.py (hunk 6 above)
Comment (mapazarr) [inline gammapy/datasets/make.py]: Fixed.

File: gammapy/datasets/make.py (hunk 7 above)
Comment (mapazarr) [inline gammapy/datasets/make.py]: It is already done... You commented on the old implementation.

File: gammapy/datasets/make.py (hunk 8 above)
Comment (mapazarr) [inline gammapy/datasets/make.py]: Dots. Fixed.

File: gammapy/scripts/make_bg_cube_models.py (hunk 9 above)
Comment (mapazarr) [inline gammapy/scripts/make_bg_cube_models.py]: Done. Maybe we need an issue 'support windows'? ;-)

File: examples/make_bg_cube_models_true-reco_test.py (hunk 10 above)
Comment (mapazarr) [inline examples/make_bg_cube_models_true-reco_test.py]: Done.

File: CHANGES.rst (hunk 11 above)
Comment (mapazarr) [inline CHANGES.rst]: Done.

File: examples/make_bg_cube_models_true-reco_test.py (hunk 12 above)
Comment (mapazarr) [inline examples/make_bg_cube_models_true-reco_test.py]: Done.

File: gammapy/scripts/make_bg_cube_models.py (hunk 9 above)
Comment (cdeil) [inline gammapy/scripts/make_bg_cube_models.py]: Oh ... we will support Windows!
And, to avoid the appearance of bias, we shall support Apple watch and MS DOS as well!

File: examples/make_bg_cube_models_true-reco_test.py (hunk 13 above)
Comment (mapazarr) [inline examples/make_bg_cube_models_true-reco_test.py]: Done.

File: examples/make_bg_cube_models_true-reco_test.py (hunk 14 above)
Comment (mapazarr) [inline examples/make_bg_cube_models_true-reco_test.py]: Done.

**Comment by cdeil** [comment]
@mapazarr – I'm going to bed in 10 min. Do you have any question here or are you all set?

Feel free to merge if you think this is ready or let me know if you want me to have another look tomorrow morning ...

File: examples/plot_bg_cube_model_comparison.py (hunk 15 above)
Comment (mapazarr) [inline examples/plot_bg_cube_model_comparison.py]: Done.

File: gammapy/datasets/make.py (hunk 16 above)
Comment (mapazarr) [inline gammapy/datasets/make.py]: Fixed.

File: gammapy/background/make.py (hunk 19 above)
Comment (mapazarr) [inline gammapy/background/make.py]: Well, actually the whole idea is to be **LESS** flexible, since we are enforcing some particular units.

In addition, the `michi` method should always use these units, since we want it to be as close to @mimayer's results as possible, in order to compare, so I think is is fine how it is.

The user can always unset this forcing of the units, by setting the varaible to `True`, in which case, the same units as the ones defined by the binning will be used (the most natural choice). If he wants something else, he will have to do it manually.

So, I'm leaving this as it is.

**Comment by mapazarr** [comment]
Ok, I addressed all comments, and travis-ci is green, so I'm merging this now.

**Comment by mapazarr** [comment]
@cdeil I forgot one last thing. I'd like to add the plto comparing tru/reco:

bg_cube_model_comparison_group27_sim_improved_true_method

to the high-level docs, at the end of:

http://gammapy.readthedocs.org/en/latest/background/make_models.html#datasets-for-testing

I think it would help illustrate the capabilities of bg cube production. The file is 260 kB; is it ok, or is it too big?

**Comment by cdeil** [comment]
@mapazarr – That is a nice Figure and it would be good to have in the docs.

But indeed, I'd prefer to keep the `gammapy` repo small and to have all example datasets and images such as this one in `gammapy-extra`. So please put this on the list of things to do for background modeling.

If you like, you can already put the image here for now:
https://github.com/gammapy/gammapy-extra/tree/master/figures
and then we'll properly include it in the Sphinx build later, when `gammapy-extra` is integrated better.

**Comment by mapazarr** [comment]
Ok, adressing the figure issue in #347.
